File: genealogy-web/src/api/tree.ts
import { apiGet } from '@/utils/request'
import type { TreeGraph } from '@/types'

export type PersonTreeDirection = 'center' | 'ancestors' | 'descendants' | 'patrilineal'

export function fetchFullTree(family_id: number) {
  return apiGet<TreeGraph>('/api/tree/full', { family_id })
}

/** 指定人物查谱（统一接口） */
export function fetchPersonTree(params: {
  family_id: number
  person_id: number
  direction: PersonTreeDirection
  up_generations?: number
  down_generations?: number
}) {
  return apiGet<TreeGraph>('/api/tree/person', params)
}

export function fetchAncestorsTree(params: {
  family_id: number
  start_generation?: number
  max_generations?: number
  person_id?: number
}) {
  return apiGet<TreeGraph>('/api/tree/ancestors', params)
}

export function fetchDescendantsTree(params: {
  family_id: number
  start_generation?: number
  max_generations?: number
  person_id?: number
}) {
  return apiGet<TreeGraph>('/api/tree/descendants', params)
}

export function fetchPatrilinealTree(params: {
  family_id: number
  root_person_id?: number
  max_generations?: number
}) {
  return apiGet<TreeGraph>('/api/tree/patrilineal', params)
}

export function fetchLineageTree(params: {
  family_id: number
  person_id: number
  up_generations?: number
  down_generations?: number
}) {
  return apiGet<TreeGraph>('/api/tree/lineage', params)
}

function mergeTreeGraphs(
  ancestors: TreeGraph,
  descendants: TreeGraph,
  focusPersonId: number,
): TreeGraph {
  const nodeMap = new Map(ancestors.nodes.map((node) => [node.id, node]))
  for (const node of descendants.nodes) {
    nodeMap.set(node.id, node)
  }

  const edgeKeys = new Set<string>()
  const edges = [...ancestors.edges, ...descendants.edges].filter((edge) => {
    const key = `${edge.source}|${edge.target}|${edge.relation}`
    if (edgeKeys.has(key)) return false
    edgeKeys.add(key)
    return true
  })

  const generations = [...nodeMap.values()]
    .map((node) => node.generation)
    .filter((generation): generation is number => generation != null)

  return {
    nodes: [...nodeMap.values()],
    edges,
    root_id: String(focusPersonId),
    is_forest: false,
    start_generation: ancestors.start_generation ?? descendants.start_generation ?? null,
    max_generation: generations.length ? Math.max(...generations) : null,
    focus_person_id: String(focusPersonId),
  }
}

function isRouteMissingError(error: unknown): boolean {
  if (!(error instanceof Error)) return false
  const message = error.message.toLowerCase()
  return message.includes('not found') || message.includes('404') || message.includes('接口不存在')
}

/** 以人为中心：优先统一接口，兼容旧后端 */
export async function fetchCenterPersonTree(params: {
  family_id: number
  person_id: number
  start_generation?: number
  up_generations?: number
  down_generations?: number
}) {
  const personParams = {
    family_id: params.family_id,
    person_id: params.person_id,
    direction: 'center' as const,
    up_generations: params.up_generations,
    down_generations: params.down_generations,
  }

  try {
    return await fetchPersonTree(personParams)
  } catch (error) {
    if (!isRouteMissingError(error)) throw error
  }

  try {
    return await fetchLineageTree(params)
  } catch (error) {
    if (!isRouteMissingError(error)) throw error
  }

  const startGeneration = params.start_generation ?? 1
  const [ancestors, descendants] = await Promise.all([
    fetchAncestorsTree({
      family_id: params.family_id,
      person_id: params.person_id,
      start_generation: startGeneration,
      max_generations: params.up_generations ?? 5,
    }),
    fetchDescendantsTree({
      family_id: params.family_id,
      person_id: params.person_id,
      start_generation: startGeneration,
      max_generations: params.down_generations ?? 5,
    }),
  ])

  return mergeTreeGraphs(ancestors, descendants, params.person_id)
}

export function formatPersonDisplayName(person: {
  name: string
  nickname?: string | null
  birth_year?: number | null
  generation?: number | null
}) {
  const base = person.nickname ? `${person.name}（${person.nickname}）` : person.name
  const meta: string[] = []
  if (person.birth_year) meta.push(String(person.birth_year))
  if (person.generation) meta.push(`第${person.generation}代`)
  return meta.length ? `${base}\n${meta.join(' · ')}` : base
}

export function modeToDirection(
  mode: 'patrilineal' | 'lineage' | 'ancestors' | 'descendants',
): PersonTreeDirection {
  if (mode === 'lineage') return 'center'
  return mode
}

File: genealogy-server/app/api/tree.py
from fastapi import APIRouter, Query

from app.core.deps import AdminUserDep, DbDep
from app.services import tree_service
from app.utils.response import success

router = APIRouter(prefix="/api/tree", tags=["tree"])


@router.get("/full")
def get_full_tree(
    db: DbDep,
    current_user: AdminUserDep,
    family_id: int = Query(...),
):
    graph = tree_service.get_full_tree(db, family_id, current_user)
    return success(graph.model_dump())


@router.get("/person")
def get_person_tree(
    db: DbDep,
    current_user: AdminUserDep,
    family_id: int = Query(...),
    person_id: int = Query(..., description="指定人物 ID"),
    direction: str = Query(
        default="center",
        description="center=以人为中心, ancestors=向上查祖, descendants=向下查孙, patrilineal=男系世系",
    ),
    up_generations: int = Query(default=5, ge=0, le=30),
    down_generations: int = Query(default=5, ge=0, le=30),
):
    graph = tree_service.get_person_tree(
        db,
        family_id,
        current_user,
        person_id=person_id,
        direction=direction,
        up_generations=up_generations,
        down_generations=down_generations,
    )
    return success(graph.model_dump())


@router.get("/ancestors")
def get_ancestors_tree(
    db: DbDep,
    current_user: AdminUserDep,
    family_id: int = Query(...),
    start_generation: int | None = Query(default=None, ge=1),
    max_generations: int = Query(default=10, ge=1, le=30),
    person_id: int | None = Query(default=None),
):
    graph = tree_service.get_ancestors_tree(
        db,
        family_id,
        current_user,
        start_generation=start_generation or 1,
        max_generations=max_generations,
        person_id=person_id,
    )
    return success(graph.model_dump())


@router.get("/patrilineal")
def get_patrilineal_tree(
    db: DbDep,
    current_user: AdminUserDep,
    family_id: int = Query(...),
    root_person_id: int | None = Query(default=None),
    max_generations: int = Query(default=12, ge=1, le=30),
):
    graph = tree_service.get_patrilineal_tree(
        db,
        family_id,
        current_user,
        root_person_id=root_person_id,
        max_generations=max_generations,
    )
    return success(graph.model_dump())


@router.get("/lineage")
def get_lineage_tree(
    db: DbDep,
    current_user: AdminUserDep,
    family_id: int = Query(...),
    person_id: int = Query(...),
    up_generations: int = Query(default=5, ge=0, le=30),
    down_generations: int = Query(default=5, ge=0, le=30),
):
    graph = tree_service.get_lineage_tree(
        db,
        family_id,
        current_user,
        person_id=person_id,
        up_generations=up_generations,
        down_generations=down_generations,
    )
    return success(graph.model_dump())


@router.get("/descendants")
def get_descendants_tree(
    db: DbDep,
    current_user: AdminUserDep,
    family_id: int = Query(...),
    start_generation: int | None = Query(default=None, ge=1),
    max_generations: int = Query(default=10, ge=1, le=30),
    person_id: int | None = Query(default=None),
):
    graph = tree_service.get_descendants_tree(
        db,
        family_id,
        current_user,
        start_generation=start_generation or 1,
        max_generations=max_generations,
        person_id=person_id,
    )
    return success(graph.model_dump())
